File: kotlin/app/src/main/java/com/vitalcheck/app/domain/entity/VitalSign.kt
package com.vitalcheck.app.domain.entity

/**
 * Entidade imutável representando uma leitura de sinais vitais.
 *
 * Utiliza data class do Kotlin para imutabilidade nativa (val),
 * equals/hashCode automáticos, e copy() para transformações seguras.
 *
 * Validação fail-fast no companion object factory — impede criação
 * de instâncias inválidas em qualquer camada.
 *
 * @property heartRate Frequência cardíaca em bpm (0–300)
 * @property steps Contagem acumulada de passos (≥ 0)
 * @property timestamp Momento da leitura
 */
data class VitalSign(
    val heartRate: Int,
    val steps: Int,
    val timestamp: Long = System.currentTimeMillis()
) {
    init {
        require(heartRate in 0..300) {
            "Frequência cardíaca inválida: $heartRate. Deve estar entre 0 e 300 bpm."
        }
        require(steps >= 0) {
            "Contagem de passos inválida: $steps. Não pode ser negativa."
        }
    }
}

File: react-native/src/domain/entities/VitalSign.ts
/**
 * Entidade que representa uma leitura de sinais vitais.
 *
 * Imutável por design — cada leitura é um snapshot no tempo.
 * Os campos são readonly para garantir imutabilidade em nível de tipo.
 */
export interface VitalSign {
  readonly heartRate: number; // bpm
  readonly steps: number;
  readonly timestamp: Date;
}

/**
 * Factory function para criação de VitalSign com validação.
 * Centraliza a lógica de criação e garante invariantes do domínio.
 */
export function createVitalSign(
  heartRate: number,
  steps: number,
  timestamp: Date = new Date()
): VitalSign {
  if (heartRate < 0 || heartRate > 300) {
    throw new Error(`Frequência cardíaca inválida: ${heartRate}. Deve estar entre 0 e 300 bpm.`);
  }

  if (steps < 0) {
    throw new Error(`Contagem de passos inválida: ${steps}. Não pode ser negativa.`);
  }

  return Object.freeze({ heartRate, steps, timestamp });
}

